tags/isp-0.12/man/ispprogress.1
.\" Copyright (C) 2005 The Regents of the University of California.
.\" Produced at Lawrence Livermore National Laboratory (cf, DISCLAIMER).
.\" Written by Jim Garlick <garlick@llnl.gov>.
.\"
.\" This file is part of ISP, a toolkit for constructing pipeline applications.
.\" For details, see <http://isp.sourceforge.net>.
.\"
.\" ISP is free software; you can redistribute it and/or modify it under
.\" the terms of the GNU General Public License as published by the Free
.\" Software Foundation; either version 2 of the License, or (at your option)
.\" any later version.
.\"
.\" ISP is distributed in the hope that it will be useful, but WITHOUT ANY
.\" WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
.\" FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
.\" details.
.\"
.\" You should have received a copy of the GNU General Public License along
.\" with ISP; if not, write to the Free Software Foundation, Inc.,
.\" 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA.
.TH ISPPROGRESS 1  2005-12-08 "" "Industrial Strength Pipes"
.SH NAME
ispprogress \- display progress bar
.SH SYNOPSIS
.BI "ispprogress -n numunits [-l string]"
.SH DESCRIPTION
\fBispprogress\fR counts the number of units passing from standard input to
standard output and displays an ASCII progress bar on standard error showing
percent complete relative to the number passed in on the command line.
.SH OPTIONS
.TP
\fB-n\fR, \fB--numunits\fR
Specify the expected number of units, used for computing the 
percentage complete.
.TP
\fB-l\fR, \fB--label\fR
Provide a label to display before the progress bar, otherwise ``progress''
will be used.
.SH CAVEATS
The progress bar will get messed up if other filters send output to standard
error.
.SH "SEE ALSO"
.BR ispbarrier (1)
.BR ispcat (1)
.BR ispexec (1)
.BR isprename (1)
.BR isprun (1)
.BR ispstats (1)
.BR ispunit (1)
.BR ispunitsplit (1)


ChangeLog
2007-06-17 Jim Garlick <garlick@llnl.gov>

 * src/Makefile: Add -fPIC to CFLAGS.

2006-02-16 Jim Garlick <garlick@llnl.gov>

 * isp.spec: use macros instead of hard coded directories.

 * : tag isp-0-10

2006-01-23 Jim Garlick <garlick@llnl.gov>

 * */*: updated headers with UC release number.

2005-12-13 Jim Garlick <garlick@llnl.gov>

 * */*: added standard copyright headers.

 * COPYING, DISCLAIMER: New.

2005-12-08 Jim Garlick <garlick@llnl.gov>

 * test/delay.c: Removed.

 * utils/ispdelay.c: New.

 * utils/{ispprogress.c,progress.[ch]}: New.

 * utils/ispcount.c: New.

 * man/*.1: Added man pages for utils.

 * utils/*: Change references to 'tag' to 'key'.

2005-12-07 Jim Garlick <garlick@llnl.gov>

 * isp/util.[ch]: make data arg of util_read and util_write (void *).

 * utils/unit.c: check string to number conversions for under/overflow.

 * utils/*: fix termination of longopts arrays.

 * utils/unit.c: added -n option.

 * utils/isprun.c: added -f option.

 * htdocs/*: New.

2005-12-05 Jim Garlick <garlick@llnl.gov> 
 * isp.spec, META: New.

 * hello/{Makefile,hello.c}: New

 * man/*: Updated.

2005-12-05 Jim Garlick <garlick@llnl.gov>

 * isp/ispunitsplit.c: fix improperly constructed isp_meta_source("split").

 * doc/{Considerations,NOTES,Overview,TODO,Proposal}: removed.

 * doc/: moved to man/.

 * doc/{report.bib,report.tex,Makefile}: added new docs.

 * etc/: removed

 * utils/isprename.c: remove support for renaming multiple files.  Fix stab
 code.

 * isp/{isp.h, error.c}: added ISP_BIND error.

 * isp/init.c: added code to check stab and display informative errors when
 binding fails.

2005-11-28 Jim Garlick <garlick@llnl.gov>

 * isp/macros.h: gcc 3.2.3 doesn't like 'default' visibility for PUBLIC 
 macro.  Make PUBLIC an empty def.

 * utils/ispunit.c: add some missing newlines in fprintf error strings.

 * isp/unit.c: renamed ISP_DOUBLE48 to ISP_DOUBLE and updated conversions.

2005-05-10 Jim Garlick <garlick@llnl.gov>

 * isp/xml_private.h: incorporated into xml.c after adding 
 xml_el_parent_get/xml_el_parent_set accessors.

 * utils/isprename.c: added symbol table.

2005-05-11 Jim Garlick <garlick@llnl.gov>

 * isp/*: API: move splitfactor argument from isp_unit_fini() to isp_init().
 Move splitfactor from result element to filter element.

 * isp/isp.c: store processed init element and provide accessors
 isp_init_get(), isp_init_set().

 * isp/init.c: added filter accessors isp_init_peek(), isp_init_find().

 * isp/filter.c: added accessors isp_filter_fid_get() and 
 isp_filter_splitfactor_get()

 * utils/ispstats.c: update for new splitfactor interface.

 * isp/init.c: added isp_init_read(), isp_init_write() functions.

 * utils/isprun.c: restored to working order.

 * test/delay.c: New.

2005-05-10 Jim Garlick <garlick@llnl.gov>

 * isp/*: replace isp_* global variables with accessor functions. 
 Misc cleanup.

2005-05-09 Jim Garlick <garlick@llnl.gov>

 * isp/isp.c: split into error.c, handle.c, init.c, isp.c, unit.c.

 * isp/macros.h: New.  

 * isp/*: added PUBLIC and PRIVATE __attribute__ macros for DSO symbol 
 management.

 * isp/isp_private.c: added new definitions for files called between library
 modules.

 * isp/{isp.[ch],file.c,unit.c}: removed result and file processing from 
 unit_read/unit_write functions and created unit_init/unit_fini.

 * isp/{isp.h, util.h}: added extern "C" { } around headers.

2005-05-06 Jim Garlick <garlick@llnl.gov>
 
 * isp/isp.[ch]: API: added 'stab' argument to isp_init and
 isp_stab_create/isp_stab_set/isp_stab_destroy functions.

 * isp/isp.[ch]: API: changed to typed metadata scheme

 * utils/*.[ch]: updated for API changes

2005-05-05 Jim Garlick <garlick@llnl.gov>

 * slurm/server.c: fixed cbuf_printf bug introduced in earlier cleanup
 that was corrupting responses.

 * slurm: split off to slurmrun project.

 * expat_test: split off to expat_test project.

 * : tag isp-0-9-prettygood

2005-05-04 Jim Garlick <garlick@llnl.gov>

 * utils/isppar.c: switch srun to unbuffered mode.  slurm's code
 for line buffering can silently corrupt data.

 * test/{test9.sh,test10,sh}: New.

 * test/sinkxml.c: check integrity of everything received.

 * test/isp.c: update file size at isp_unit_write() time, not earlier.
 verify that size is correct even if we are not verifying md5.  Make size 
 unsigned long bytes not int megabytes.

 * isp/util.c: util_poll: restored infinite timeout for poll when tv is NULL.
 Everything was spinning.

 * test/{srcxml.c,sinkxml.c}: added test for excessive polling.

 * slurm/isprun.c: renamed to ispsrun.c

 * utils/isppar.c: renamed to isprun.c

 * utils/ispsplit.c: renamed to ispunitsplit.c

2005-05-03 Jim Garlick <garlick@llnl.gov>
 
 * isp/isp.[ch]: added ISP_EELEMENT and ISP_EDOCUMENT errors.
 Removed ISP_EINIT error.

 * isp/isp.c: removed isp_proxy_flag.  Use handle->flags instead.

 * utils/isppar.c: fixed to reflect updates in (private) API.

2005-05-02 Jim Garlick <garlick@llnl.gov>

 * isp/{xin.h,xout.h: inline documentation

 * isp/{isp_private.h, isp.c}: changed isp_init_sink()/isp_init_source() to
 isp_read_init()/isp_write_init().

 * isp/isp.c: cleaned up isp_init().

2005-04-29 Jim Garlick <garlick@llnl.gov>

 * isp/xout.[ch]: makeover.  Switch from cbuf to internal element buffering
 with configurable backlog.

 * isp/cbuf.[ch]: removed.

 * isp/xin.[ch]: added backlog limit.

 * isp/isp.[ch]: updated for xout/xin changes.  Added return value to
 isp_fini() and isp_handle_destroy() since they can return I/O errors when
 the xout buffer is flushed on destruction.

 * utils/*, test/*: updated for API changes.

2005-04-28 Jim Garlick <garlick@llnl.gov>

 * isp/xout.c: Remove internal XML document cache.  Do not free element 
   in xout_write_el() (caller must now free).

 * isp/isp.[ch]: API: split up _unit_write().  As a result of 
 xout_write_el() change, caller must now free unit after isp_unit_write().

 * utils/*: update for API cleanup and added isp_unit_destroy() calls.


isp.old/utils/ispprogress.c
/*****************************************************************************\
 *  $Id$
 *****************************************************************************
 *  Copyright (C) 2005 The Regents of the University of California.
 *  Produced at Lawrence Livermore National Laboratory (cf, DISCLAIMER).
 *  Written by Jim Garlick <garlick@llnl.gov>.
 *  
 *  This file is part of ISP, a toolkit for constructing pipeline applications.
 *  For details, see <http://isp.sourceforge.net>.

 *  ISP is free software; you can redistribute it and/or modify it under
 *  the terms of the GNU General Public License as published by the Free
 *  Software Foundation; either version 2 of the License, or (at your option)
 *  any later version.
 *  
 *  ISP is distributed in the hope that it will be useful, but WITHOUT ANY
 *  WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 *  FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 *  details.
 *  
 *  You should have received a copy of the GNU General Public License along
 *  with ISP; if not, write to the Free Software Foundation, Inc.,
 *  59 Temple Place, Suite 330, Boston, MA  02111-1307  USA.
\*****************************************************************************/

/* Given a number of "expected" units, show an ascii progress bar
 * on stderr indicating percentage complete.
 */

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif
#define _GNU_SOURCE
#include <stdio.h>
#include <string.h>
#include <getopt.h>
#include <stdlib.h>
#include <limits.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <assert.h>
#include <dirent.h>
#include <errno.h>
#include <math.h>

#include <isp/isp.h>
#include "progress.h"

#define OPT_STRING "n:l:"
static const struct option long_options[] = {
    {"numunits", required_argument, 0, 'n'},
    {"label", required_argument, 0, 'l'},
    {0,0,0,0},
};
static const struct option *longopts = long_options;

static char *progname = NULL;
static prog_t pctx;
static char *label = "progress";

static void 
usage(void)
{
    fprintf(stderr, "Usage: %s -n numunits [-l label]\n", progname);
    exit(1);
}

static int
progress(isp_unit_t u, void *arg)
{
    unsigned long numunits = *(unsigned long *)arg;
    static unsigned long count = 0;

    count++;
    progress_update(pctx, (double)count/numunits);

    return ISP_ESUCCESS;
}

int 
main(int argc, char *argv[])
{
    int c;
    int longindex;
    int res;
    isp_handle_t h;
    unsigned long numunits = 0;
   
    opterr = 0;
    progname = basename(argv[0]);

    while ((c = getopt_long(argc, argv, OPT_STRING, longopts, 
                    &longindex)) != -1) { 
        switch (c) { 
            case 'n':   /* --numunits */
                numunits = strtoul(optarg, NULL, 10);
                if (numunits == ULONG_MAX && errno == ERANGE) {
                    fprintf(stderr, "%s: numunits value overflow", progname);
                    exit(1);
                }
                break;
            case 'l':   /* --label */
                label = optarg;
                break;
            default:
                usage();
                /*NOTREACHED*/
        }
    }
    if (numunits == 0) 
        usage();
    if (optind < argc)
        usage();

    res = isp_init(&h, ISP_SOURCE|ISP_SINK, argc, argv, NULL, 1);
    if (res != ISP_ESUCCESS)
        isp_errx(1, "isp_init: %s", isp_errstr(res));

    fprintf(stderr,"%s: ", label);
    progress_create(&pctx, 79 - strlen(label) - 2);

    if ((res = isp_unit_map(h, progress, &numunits)) != ISP_ESUCCESS)
        isp_errx(1, "isp_map: %s", isp_errstr(res));

    progress_destroy(pctx);

    if ((res = isp_fini(h)) != ISP_ESUCCESS)
        isp_errx(1, "isp_fini: %s", isp_errstr(res));

    exit(0);
}

/*
 * vi:tabstop=4 shiftwidth=4 expandtab
 */
